### PR TITLE
Actionable error message for large json documents

### PR DESCRIPTION
**Pull request by hibikiledo**
### Problem

When using `assertJsonDocumentMatches`, it is hard to navigate the error message when the assertion fails.

The message only states that the given JSON document is not a match of the provided constraints.
It would be great to also tell _which constraint(s)_ the JSON document fails to satisfy.

### Solution

I implement `additionalFailureDescription` in `JsonValueMatchesMany` constraint to return a string representation of constraints that fail during evaluations.

The result would look like this.

Screen Shot 2564-03-21 at 07 41 33

I think this would greatly simplify investigation when the tests fail.

Thank you

### CODE REVIEW

File: src/Constraint/JsonValueMatchesMany.php [hunk 1]
```diff
@@ -59,11 +62,23 @@ function (Constraint $constraint) {
      */
     protected function matches($other): bool
     {
+        $result = true;
         foreach ($this->constraints as $constraint) {
             if (!$constraint->evaluate($other, '', true)) {
-                return false;
+                $result = false;
+                $this->failedConstraints[] = $constraint->toString();
```
Comment (martin-helmich) [inline src/Constraint/JsonValueMatchesMany.php:69]: This introduces a state to the class that will persist across multiple invocations. This is most probably an edge case, but _will_ lead to odd error messages when the same instance of a `JsonValueMatchesMany` constraint is re-used multiple times across a test case.

When looking at other implementations of `additionalFailureDescription` (PHPUnit itself uses it only once, in the `StringMatchesFormatDescription` constraint), these are stateless (i.e. re-compute the failure description there and then based on the `$other` input).

https://github.com/sebastianbergmann/phpunit/blob/99f6e264b464b3a9089ea2dceed79c3d953244e8/src/Framework/Constraint/String/StringMatchesFormatDescription.php#L58-L77

**Review by martin-helmich — COMMENTED** [review]
Thanks for the contribution, @hibikiledo! :+1: Full ACK that this is an entirely sensible change to make -- the validation messages, especially for larger assertions, are currently not optimal and can be improved significantly.

I'm just having one thought regarding the statefulness of the `JsonValueMatchesAny` constraint (see the respective comment). Let me know your thoughts on this.

File: src/Constraint/JsonValueMatchesMany.php (hunk 1 above)
Comment (hibikiledo) [reply src/Constraint/JsonValueMatchesMany.php:69]: Thanks for the review.

I didn't look at other implementations before attempting to work on this.
Next time, I will make sure to explore further before jumping in. 💯

I think it is better to keep the class stateless.
Stateless class is simple and easy to reason about and also more resilient to changes.

I will update the PR shortly. 👍

Comment (hibikiledo) [reply src/Constraint/JsonValueMatchesMany.php]: I have updated the `additionalFailureDescription` method to build message statelessly.

Please kindly review again.

File: src/Constraint/JsonValueMatchesMany.php [hunk 2]
```diff
@@ -66,4 +66,23 @@ protected function matches($other): bool
         }
         return true;
     }
+
+    /**
+     * Returns a string representation of matches that evaluate to false.
+     *
```
Comment (martin-helmich) [inline src/Constraint/JsonValueMatchesMany.php:72]: This will make codeclimate stop complaining:

```suggestion
     *
```

**Review by martin-helmich — CHANGES_REQUESTED** [review]
:+1: Looking good! I just added (and already applied) one quick suggestion (whitespace only) to make codeclimate stop complaining.

**Review by martin-helmich — APPROVED** [review]
